File: java-to-python-practice/valid-parentheses/validParentheses.java
class Solution {
    public boolean isValid(String string) {
        Map<Character,Character> mappings = new HashMap<Character,Character>(){{
            put(')', '(');
            put('}', '{');
            put(']', '[');
        }};
        
        Stack<Character> unclosedParens = new Stack();
        
        for (int i = 0; i < string.length(); i++) {
            char curChar = string.charAt(i);
            
            boolean isClosingParen = mappings.containsKey(curChar);
            if (isClosingParen) {
                char matchingOpenParen = mappings.get(curChar);
                if (!unclosedParens.isEmpty() && unclosedParens.pop() == matchingOpenParen) continue;
                
                return false;
            }
            else {
                unclosedParens.push(curChar);
            }
        }
        
        return unclosedParens.isEmpty();
    }
}

File: java-to-python-practice/valid-parentheses/valid_parentheses.py
class Solution:
    def isValid(self, string: str) -> bool:
        mappings = {
            ')': '(',
            '}': '{',
            ']': '['
        }
        
        unclosed_parens = []
        
        for char in string:
            is_closing_paren = char in mappings
            
            if is_closing_paren:
                matching_open_paren = mappings[char]
                if len(unclosed_parens) > 0 and unclosed_parens.pop() == matching_open_paren:
                    continue
                    
                return False
            else:
                unclosed_parens.append(char)
                
        return len(unclosed_parens) == 0
    
